UI/MapUI.gd

```diff
@@ -18,6 +24,59 @@ func _input(_inputevent: InputEvent) -> void:
 
 func _on_return_button_press():
 	queue_free()
+	
+
+func _on_add_torch_pressed():
+	torch_confirmation_dialog.show()
+
+func _close_torch_placement_dialog():
+	torch_confirmation_dialog.hide()
+
+func _add_torch_to_current_location():
+	var current_player_floor_index = current_player_room.floor_index
+	var current_player_room_index = current_player_room.room_index
+	
+	var room_light_range = 1
+	var max_light_index = current_player_floor_index + _LIGHT_FLOOR_RANGE + 1 
+	
+	if current_player_floor_index + _LIGHT_FLOOR_RANGE + 1 > MapManager.map_width_array.size():
+		max_light_index = MapManager.map_width_array.size() 
+	
+	for floor_light_index: int in range(current_player_floor_index + 1, max_light_index):
```

Comment:
I feel like all of what is done after in this function is setting level level in rooms, which i feel should be done in `_add_light_to_rooms` ?
I would also have done this recursively instead of a for loop. Start with `remaining_light_range` equal to the how far the light goes from a torch (so 3 for us, idk which const does that exactly).
Look at the rooms accessibly by the player and then two things happen:
- the remaining light range is 1, change the light level and return
- the remaining light range is not 1, apply this function to all accessible rooms and change the light level of the room you are looking at

For me this is much more intuitive to look at than for loops with carefully chosen bounds at each step, which is very prone to index errors also

Reply:
i'm realizing you will also need to have a boolean to know if a room had its light changed already by another function call during this light change.
Because you could access the same room in multiple ways, like going left-right or right-left will lead you to the same room (when i say right I mean up-right in diagonal, we don't stay on the same floor)
so something like that:
```GDScript
func _add_light_to_rooms(floor_index: int = current_player_room.floor_index, room_index: int = current_player_room.room_index, remaining_light_range: int = TOTAL_LIGHT_RANGE, is_room_already_changed: bool = false) -> bool:
	if not is_room_already_changed:
		if remaining_light_range == 1:
			# change the light
			return true
		else:
			var accessible_rooms: Array[RoomBase] = get_all_accessible_rooms # to implement
			for current_room: RoomBase (RoomUI ?)  in accessible_rooms:
				_add_light_to_rooms(current_room.floor_index, current_room.room_index, remaining_light_range -1, false)
			# change the light in the current room
	return true
```
and writing this noticed that a boolean wouldn't work :laughing: maybe returning a list of all the rooms that are already changed (by returning a hash of the room for example, or an object ID when changing the light), and checking that the `current_room` is not in the list of already changed rooms ?

Reply:
Everything in `_add_torch_to_current_location()` refactored to use `MapMovement.get_all_accessible_rooms_in_range()`